pkg/threescale/backend/backend.go

```diff
@@ -162,6 +162,61 @@ out:
 	return commitChanges
 }
 
+// Report the provided request transactions to the cache
+// Report always returns with 'Accepted' as bool 'true' and the error will only ever be nil
+// in cases where  the provided transaction are empty or nil
+// The background task will abort in cases where we have a cache miss and we also fail to get the hierarchy from 3scale
+// Typically, we can handle cache miss and report to cache once we are aware of a hierarchy for the service
+// Request Transactions must not be nil and must not be empty
+// Supports multiple transactions
+func (b *Backend) Report(request threescale.Request) (*threescale.ReportResult, error) {
+	return b.report(request)
+}
+
+func (b *Backend) report(request threescale.Request) (*threescale.ReportResult, error) {
+	var hierarchy api.Hierarchy
+	var err error
+
+	err = validateTransactions(request.Transactions)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for _, transaction := range request.Transactions {
+			// we support reporting in batches so for every transaction, grab the cache key and see if we
+			// have a match. if not, we can report locally regardless once we know the hierarchy
+			cacheKey := generateCacheKeyFromRequest(request)
```

Comment:
How is this dependent on the transaction yielded by the iterator? Since it only depends on the request, which is constant in this context, `cacheKey` can be extracted out of the loop. But that would not be correct, because each transaction should refer to a (usually) different application. The cache key definitely needs to depend on the exact transaction being used, and the issue is that this func is only accounting for the first transaction in the request.

We can fix this by accepting a transaction number or pointer as second parameter to `generateCacheKeyFromRequest`, and use that rather than just the first transaction.